src/physics/em/detail/BetheBlochInteractor.i.hh

```diff
@@ -0,0 +1,203 @@
+    }
+
+    real_type kinetic_energy = inc_energy_.value();
+    
+    real_type tmin = min(kinetic_energy, this->min_incident_energy().value());
+
+    ElementView element = material_.element_view(celeritas::ElementComponentId{0});
+
+    real_type func1 = tmin * this->differential_cross_section(tmin, element);
+
+    real_type ln_epsilon, epsilon;
+    real_type func2;
+
+    real_type xmin = std::log(tmin);
+    real_type xmax = std::log(kinetic_energy / tmin);
+    UniformRealDistribution<real_type> p;
+
+    do
+    {
+        ln_epsilon = xmin + p(rng) * xmax;
+        epsilon = std::exp(ln_epsilon);
+        func2 = epsilon * this->differential_cross_section(epsilon, element);
+    } while(func2 < func1 * p(rng));
+
+    real_type gamma_energy = epsilon;
+
+    // Sample secondary direction.
+    UniformRealDistribution<real_type> phi(0, 2 * constants::pi);
+
+    real_type cost = this->sample_cos_theta(gamma_energy, rng);
+    Real3 gamma_dir = rotate(from_spherical(cost, phi(rng)), inc_direction_);
+
+    real_type tot_momentum = std::sqrt(gamma_energy *
+                                    (gamma_energy + 2.0 * inc_mass_.value()));
+    Real3 inc_direction;
+    for (int i = 0; i < 3; i++)
+    {
+        inc_direction[i] = tot_momentum * inc_direction_[i]
+                                        - gamma_energy * gamma_dir[i];
+    }
+    normalize_direction(&inc_direction);
+
+    // Construct interaction for change to primary (incident) particle
+    Interaction result;
+    result.action      = Action::scattered;
+    result.energy      = units::MevEnergy{kinetic_energy - gamma_energy};
+    result.direction   = inc_direction;
+    result.secondaries = {secondaries, 1};
+
+    // Save outgoing secondary data
+    secondaries[0].particle_id = shared_.gamma_id;
+    secondaries[0].energy      = units::MevEnergy{gamma_energy};
+    secondaries[0].direction   = gamma_dir;
+
+    return result;
+}
+
+template<class Engine>
+CELER_FUNCTION real_type
+BetheBlochInteractor::sample_cos_theta(real_type gamma_energy,
+                                       Engine& rng)
+{
+    real_type mass = inc_mass_.value();
+    real_type gam  = 1.0 + inc_energy_.value() / mass;
+    real_type rmax = gam * constants::pi * 0.5
+                         * min(1.0, gam * mass / gamma_energy - 1);
+
+    UniformRealDistribution<real_type> xi;
+    real_type x = xi(rng) * ipow<2>(rmax) / (1 + ipow<2>(rmax));
+
+    return std::cos(std::sqrt(x / (1.0 - x)) / gam);
+}
+
+CELER_FUNCTION real_type
+BetheBlochInteractor::differential_cross_section(real_type gamma_energy,
+                                                 ElementView element)
+{
+
+    real_type dxsection = 0.0;
+    real_type ep_max = inc_energy_.value();
+
+    if (gamma_energy >= ep_max)
+    {
+        return dxsection;
+    }
+
+    int Z = element.atomic_number();
+    real_type A = element.atomic_mass().value();
+    real_type mass = inc_mass_.value();
+    real_type sqrte = 1.648721271;
```

Comment:
We should add `e` to `src/base/Constants.hh` and calculate this quantity here as const

```c++
const real_type sqrte = std::sqrt(constants::euler);
```

so we keep full control of our constants and don't end up with `e` or `pi` hardcoded with different precisions in different parts of the code.